File: filemaid/cleaners/npm.py
"""npm cache cleanup."""
import shutil
import subprocess
from pathlib import Path

from filemaid.cleaners._result import CleanupResult
from filemaid.cleaners._size import dir_size, humanize


def _npm_cache_dir() -> Path:
    try:
        result = subprocess.run(
            ["npm", "config", "get", "cache"],
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            text=True,
            timeout=30,
        )
        path = result.stdout.strip()
        if path and path != "undefined":
            return Path(path).expanduser()
    except Exception:
        pass
    return Path.home() / ".npm"


def can_run() -> bool:
    return shutil.which("npm") is not None


def run(dry_run: bool, config: dict) -> CleanupResult:
    cache_dir = _npm_cache_dir()
    content_dir = cache_dir / "_cacache"
    before = dir_size(content_dir) if content_dir.exists() else 0
    if dry_run:
        return CleanupResult(
            name="npm",
            status="dry-run",
            saved=before,
            saved_human=humanize(before),
            detail="would clean cache",
        )
    try:
        result = subprocess.run(
            ["npm", "cache", "clean", "--force"],
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
            timeout=120,
        )
        after = dir_size(content_dir) if content_dir.exists() else 0
        saved = max(0, before - after)
        output = result.stdout.strip()
        return CleanupResult(
            name="npm",
            status="ok",
            saved=saved,
            saved_human=humanize(saved),
            detail=f"cache cleaned\n{output}" if output else "cache cleaned",
        )
    except Exception as exc:
        return CleanupResult(
            name="npm",
            status="failed",
            detail=str(exc),
        )

File: internal/cleaners/npm.go
package cleaners

import (
	"path/filepath"
	"strings"

	"github.com/logicminds/filemaid/internal/config"
)

var npmCleaner = Cleaner{
	Name:   "npm",
	CanRun: npmCanRun,
	Run:    npmRun,
}

func npmCanRun() bool {
	_, err := lookPath("npm")
	return err == nil
}

func npmCacheDir() string {
	out, err := runner.Run("npm", "config", "get", "cache")
	if err == nil {
		path := strings.TrimSpace(string(out))
		if path != "" && path != "undefined" {
			return expandPath(path)
		}
	}
	return filepath.Join(home(), ".npm")
}

func npmRun(dryRun bool, cfg *config.Config) CleanupResult {
	cacheDir := npmCacheDir()
	contentDir := filepath.Join(cacheDir, "_cacache")
	before := dirSizeIfExists(contentDir)

	if dryRun {
		saved := before
		return CleanupResult{
			Name:       "npm",
			Status:     "dry-run",
			Saved:      &saved,
			SavedHuman: Humanize(before),
			Detail:     "would clean cache",
		}
	}

	out, err := runner.Run("npm", "cache", "clean", "--force")
	if err != nil {
		return CleanupResult{
			Name:   "npm",
			Status: "failed",
			Detail: err.Error(),
		}
	}

	after := dirSizeIfExists(contentDir)
	saved := before - after
	if saved < 0 {
		saved = 0
	}
	output := strings.TrimSpace(string(out))
	detail := "cache cleaned"
	if output != "" {
		detail = "cache cleaned\n" + output
	}

	return CleanupResult{
		Name:       "npm",
		Status:     "ok",
		Saved:      &saved,
		SavedHuman: Humanize(saved),
		Detail:     detail,
	}
}
